imaginex_lambda/lib/img_lib.py

```diff
@@ -0,0 +1,197 @@
+            this width.
+        height (Optional[int]): the maximum height of the image. If the image is wider than this value, it will be resized to fit within
+            this height.
+    Returns:
+        bytes: Optimized image data
+    """
+
+    logger.info("Optimizing image...")
+
+    with ExitStack() as stack:
+        img = stack.enter_context(Image.open(buffer))
+        if width and width < img.width:
+            logger.info(f"Resizing image given width {width}px...")
+            new_height = int(width * img.height / img.width)
+            logger.info("New height: %d", new_height)
+            img = stack.enter_context(img.resize((width, new_height)))
+            logger.info("Resized image to width: %d and height: %d", width, new_height)
+        elif height and height < img.height:
+            logger.info(f"Resizing image to the given {height}px...")
+            new_width = int(height * img.width / img.height)
+            logger.info("New height: %d", new_width)
+            img = stack.enter_context(img.resize((new_width, height)))
+            logger.info("Resized image to width: %d and height: %d", height, new_width)
+        tmp = stack.enter_context(BytesIO())
+        img.save(tmp, quality=quality, optimize=True, format=ext)
+        tmp.seek(0)
+
+        logger.info("Optimized image!")
+        return tmp.read()
+
+
+def download_and_optimize(url: str,
+                          quality: int,
+                          width: Optional[int],
+                          height: Optional[int],
+                          bucket_name: str,
+                          chunk_size: int = 1024) -> Tuple[bytes, str, float]:
+    """
+    This is the function responsible for coordinating the download and optimization of the images. It should
+    not concern itself with any lambda-specific information.
+
+    Args:
+        url (str): The URL of the image to download.
+        quality (int): The quality to optimize the image (0 to 100).
+        width (Optional[int]): The width of the image to resize to.
+        height (Optional[int]): The height of the image to resize to.
+        bucket_name (str): The name of the S3 bucket to download the image from.
+        chunk_size (int): The chunk size to use when downloading the image.
+
+    Returns:
+        Tuple[bytes, str, float]: A tuple containing the optimized image data, content type, and the compression ratio.
+
+    Raises:
+        HandlerError: If `url` is empty, `width` and `height` are both empty or both provided, or if `width`
+        or `height` are less than or equal to zero.
+
+    """
+
+    if not url:
+        raise HandlerError('url is required')
+
+    if width is None and height is None:
+        raise HandlerError('Width or height must be defined')
+
+    if width is not None and height is not None:
+        raise HandlerError('Only one of width or height params must be defined')
```

Comment:
I believe that in lot of cases, Frontend has both width and height specified so it gets propagated to API call.
We should probably pick one of it based on ration. Aka. one of the values won't be respected and image will be smaller than preferred in one direction.

Example:
- I send 300px w and 300px h
- If image is landscape, I use width and height is smaller than 300px,
- If image is portrait, I use height and width is smaller than 300px.

Reply:
Good idea... let me take a look at it.